setup/lib/utils.js

```diff
@@ -0,0 +1,74 @@
+const { setFailed } = require('@actions/core');
+const os = require('os');
+const path = require('path');
+const execSync = require('child_process').execSync;
+const semver = require('semver');
+const core = require('@actions/core');
+
+// arch in [arm, x32, x64...] (https://nodejs.org/api/os.html#os_os_arch)
+// return value in [amd64, 386, arm]
+function mapArch(arch) {
+    const mappings = {
+        x32: '386',
+        x64: 'amd64'
+    };
+    return mappings[arch] || arch;
+}
+
+// os in [darwin, linux, win32...] (https://nodejs.org/api/os.html#os_os_platform)
+// return value in [darwin, linux, windows]
+function mapOS(os) {
+    const mappings = {
+        darwin: 'macOS',
+        win32: 'windows'
+    };
+    return mappings[os] || os;
+}
+
+function getDownloadObject(version) {
+    const platform = os.platform();
+    const filename = `notation_${version}_${mapOS(platform)}_${mapArch(os.arch())}`;
+    const extension = platform === 'win32' ? 'zip' : 'tar.gz';
+    const binPath = platform === 'win32' ? 'bin' : path.join(filename, 'bin');
+    const url = `https://github.com/notaryproject/notation/releases/download/v${version}/${filename}.${extension}`;
```

Comment:
As mentioned above, making `url` as an input is more flexible.

Reply:
Is the suggestion to replace `https://github.com/notaryproject/notation/releases/download/v` with the input from the `url` parameter from the action?

So the const would look like this:

```js
const downloadUrl = core.getInput('url');
   const url = `${downloadUrl}/${version}/${filename}.${extension}`;
```

And then the actions.yml could be updated to have the default of:

```yml
  url:
    description: URL of Notation CLI to install
    required: false
    default: https://github.com/notaryproject/notation/releases/download/v
```

Reply:
For user input `url`, we could make it a full valid url (without default) and add an input called `checksum` in `action.yml`. If user inputs a `url`, `version` will be skipped. The github action downloads Notation CLI from the url directly and check the checksum based on user input (we could add this logic in the refactoring PR later, this logic will be duplicated in many places). So something like this:
```
 url:
    description: URL of Notation CLI to install
    required: false
 checksum:
    description: checksum of the downloaded Notation CLI tarball. Required if `url` is provided.
    required: false
```

Reply:
> Imo, with the addition of the checksum input this work should be done under the PR for the checksum implementation with the refactor work. If I add the checksum input now there won't be any logic behind it with the current version of the code base.
> 
> Completely agree with the change though. :)

Sure. I agree with you @duffney. Please feel free to resolve the comment (somehow I'm not able to resolve it from my side).

Reply:
Issue created: https://github.com/notaryproject/notation-action/issues/11